### PR TITLE
Fix dereference value-initialized vector iterator

### PR DESCRIPTION
**Pull request by bellenot**
Fix the following error on Windows:
```
Debug assertion Failed!
Can't dereference value-initialized vector iterator
```
As the standard says:
`std::vector::end()` returns an iterator to the element following the last element of the vector.
This element acts as a placeholder; attempting to access it results in undefined behavior.
See https://en.cppreference.com/w/cpp/container/vector/end 
And with empty vectors, vector.begin() == vector.end()

### CODE REVIEW

File: io/io/src/TEmulatedMapProxy.cxx [hunk 1]
```diff
@@ -71,7 +71,7 @@ void* TEmulatedMapProxy::At(UInt_t idx)
    // Return the address of the value at index 'idx'.
    if ( fEnv && fEnv->fObject )   {
       PCont_t c = PCont_t(fEnv->fObject);
-      return idx<(c->size()/fValDiff) ? ((char*)&(*c->begin())) + idx*fValDiff : 0;
+      return idx<(c->size() / fValDiff) ? c->data() + idx * fValDiff : 0;
```
Comment (pcanal) [inline io/io/src/TEmulatedMapProxy.cxx:74]: Minor readability improvement on the existing code:
```suggestion
      return (idx<(c->size() / fValDiff)) ? (c->data() + idx * fValDiff) : 0;
```

File: io/io/src/TGenCollectionProxy.cxx [hunk 2]
```diff
@@ -1514,13 +1514,8 @@ void TGenCollectionProxy__VectorCreateIterators(void *obj, void **begin_arena, v
...
    }
-   *begin_arena = &(*vec->begin());
-#ifdef R__VISUAL_CPLUSPLUS
-   *end_arena = &(*(vec->end()-1)) + 1; // On windows we can not dererence the end iterator at all.
-#else
-   // coverity[past_the_end] Safe on other platforms
-   *end_arena = &(*vec->end());
-#endif
+   *begin_arena = vec->data();
+   *end_arena = vec->data() + vec->size(); // We can not dereference the end iterator at all.
```
Comment (pcanal) [inline io/io/src/TGenCollectionProxy.cxx:1518]:
```suggestion
   *end_arena = vec->data() + vec->size();
```
The comment seems useful only in reference to the now gone code.

**Review by pcanal — APPROVED** [review]
LVGTM.  Thank you!

File: io/io/src/TEmulatedMapProxy.cxx (hunk 1 above)
Comment (bellenot) [reply io/io/src/TEmulatedMapProxy.cxx:74]: right, I missed this one, thanks!

**Pull request merged**